sympy/solvers/ode/systems.py

```diff
@@ -0,0 +1,375 @@
+    t: Symbol
+        Independent variable of the equations in eqs
+
+    Returns
+    =======
+
+    match = {
+        'no_of_equation': len(eqs),
+        'eq': eqs,
+        'func': funcs,
+        'order': order,
+        'is_linear': is_linear,
+        'is_constant': is_constant,
+        'is_homogeneous': is_homogeneous,
+    }
+
+    Dict or None
+        Dict with values for keys:
+            1. no_of_equation: Number of equations
+            2. eq: The set of equations
+            3. func: List of dependent variables
+            4. order: A dictionary that gives the order of the
+                      dependent variable in eqs
+            5. is_linear: Boolean value indicating if the set of
+                          equations are linear or not.
+            6. is_constant: Boolean value indicating if the set of
+                          equations have constant coefficients or not.
+            7. is_homogeneous: Boolean value indicating if the set of
+                          equations are homogeneous or not.
+        This Dict is the answer returned if the eqs are linear and constant
+        coefficient. Otherwise, None is returned.
+    """
+    from sympy.solvers.solveset import linear_eq_to_matrix
+    from sympy.solvers.solvers import solve
+
+    # Error for i == 0 can be added but isn't for now
+
+    # Removing the duplicates from the list of funcs
+    # meanwhile maintaining the order. This is done
+    # since the line in classify_sysode: list(set(funcs)
+    # cause some test cases to fail when gives different
+    # results in different versions of Python.
+    funcs = list(uniq(funcs))
+
+    # Check for len(funcs) == len(eqs)
+    if len(funcs) != len(eqs):
+        raise ValueError("Number of functions given is not equal to the number of equations %s" % funcs)
+
+    # ValueError when functions have more than one arguments
+    for func in funcs:
+        if len(func.args) != 1:
+            raise ValueError("dsolve() and classify_sysode() work with "
+            "functions of one variable only, not %s" % func)
+
+    # Getting the func_dict and order using the helper
+    # function
+    order = _get_func_order(eqs, funcs)
+
+    # Not adding the check if the len(func.args) for
+    # every func in funcs is 1
+
+    rep = {func.diff(t, n): Dummy() for func in funcs for n in range(order[func] + 1)}
+    eqs_sub = [eq.subs(rep) for eq in eqs]
+
+    # Linearity check
+    try:
+        A, b = linear_eq_to_matrix(eqs_sub, rep.values())
+
+    except ValueError:
+        return None
+
+    is_linear = True
+
+    # Constant coefficient check
+    is_constant = True
+    for coef in A:
+        if is_constant == True:
+            if coef.as_independent(t, as_Add=True)[1] != 0:
+                is_constant = False
+
+    # Homogeneous check
+    is_homogeneous = True if b.is_zero_matrix else False
+
+    match = {
+        'no_of_equation': len(eqs),
+        'eq': eqs,
+        'func': funcs,
```

Comment:
We should probably pass `A` and `b` here to save recomputing them later.

Reply:
@oscarbenjamin Where would we need ```A``` and ```b``` exactly after determining whether the system is constant and homogeneous? I am saying this because we compute ```A``` and ```b``` considering the fact that ```AX = b``` is a matrix equation such that the vector ```X``` has all the dependent variables in every order that it is present in the system of ODEs. But we want something like ```X' = AX + b``` where ```X``` is a vector of dependent variables and ```X'``` is the vector of dependent variables in its first order.

Reply:
It's probably something to handle in a later PR but I would expect that the canonical form of say a system of linear 1st order ODEs would be:
```
A X' + B X = f
```
For 2nd order
```
A X'' + B X' + C X = f
```
We could make a function that returns e.g. `((A, B), f)` or `((A, B, C), f)`. Mixed order could just use the max order.

For use here in solving we would probably want to solve for the highest derivative of each function first meaning that the matrices A above would just be the identity matrix. For mixed order we would want to recast into a common order (e.g. 1st) before extracting the matrices.

Reply:
@oscarbenjamin Ok, then wouldn't we need to update the ```match``` dictionary to have keys like: ```A```, ```b```, ```rep``` since the rep has all the data about the functions in their order and the Dummy name it has been assigned?

Reply:
I guess so. Ultimately I think we should do away with this idea that we match first and then solve after. We are in the process of refactoring dsolve for single ODEs so that it uses classes for each solver that can store whatever state they need. Once we are done replacing all of the solvers we can rearrange the logic of `dsolve_system`.